fix: bind punctuation to following word when needed

cblanken opened the pull request:
This PR fixes issue #328.

The existing punctuation spacing implementation uses a `spaceAfter` attribute to force a space (right padding) after punctuation from the `tokens_with_no_space_before` config array. However, for many punctuation marks this reduces readability since punctuation (like quotation marks) may appear to be bound to the previous word when it should be bound to the next word.

For example:
> Tommy said, "Hello there".

would be incorrectly rendered like this

> Tommy said," Hello there".

This fix propagates the `is_punct` and `whitespace_` attributes from the spacy tokenization and a boolean (`space_before`) for tokens with preceding whitespace.  They're all used to more accurately arrange the space surrounding punctuation.

@simjanos-dev I made some progress on this, but I've only tested it with German so far. It seems to work well, but I was wondering if you had a particular procedure or some sample texts you used to test? I want to make sure my changes didn't break rendering for the other languages.

## New rendering
![image](https://github.com/user-attachments/assets/c2589185-7559-40e0-97e2-d3f2e638e159)

## Previous rendering
![image](https://github.com/user-attachments/assets/9167dd73-e450-4b46-86de-09560be84333)

cblanken committed:
fix: bind punctuation to following word when needed

The existing punctuation spacing implementation uses a `spaceAfter` attribute
to force a space (right padding) after punctuation from the
`tokens_with_no_space_before` config array. However, for many punctuation marks
this reduces readability since punctuation (like quotation marks) may appear to be
bound to the previous word when it should be bound to the next word.

For example:
> Tommy said, "Hello there".
would be incorrectly rendered like this
> Tommy said," Hello there".

This fix propagates the `is_punct` and `whitespace_` attributes from the spacy
tokenization and a boolean (`space_before`) for tokens with preceding
whitespace.  They're all used to more accurately arrange the space surrounding
punctuation.

simjanos-dev commented:
Hi!

Sorry for not responding for this long, did not have energy to check this out yet, then kind of forgot.

> I made some progress on this, but I've only tested it with German so far. It seems to work well, but I was wondering if you had a particular procedure or some sample texts you used to test? I want to make sure my changes didn't break rendering for the other languages.

I ususally using lingua dot com website for languages that it has. I use nhk news easy for Japanese (the button next to the sound button turns off hiragana above the words, then you can copy articles). For other languages I just google language + simple story.

If you want to test it, the current unique languages are Japanese, Chinese and Thai. The rest work mostly the same, although tokenization part is always dependent on the spacy model.

Are `"` characters in your improved rendering separate words?

cblanken commented:
Thanks, I'll do some more testing on the other languages when I get a chance.

> Are " characters in your improved rendering separate words?
Yeah, they're still separate words. I didn't change the tokens just some of the metadata for rendering.

![image](https://github.com/user-attachments/assets/454b96bd-fcda-4a1f-ae4f-de6ceea2d224)

Here's an example with them not marked as known.

cblanken committed:
fix: default to legacy text rendering when needed

Previously imported texts (those without the additional `is_punct` and
`space_before` properties) must use the original hard-coded space-after
rendering. Otherwise, no spaces are rendered between tokens.

cblanken commented:
Okay, I tested a few other languages (Chinese, Japanese, and Russian) and they seem to render fine, but Thai texts are failing to import. Will have to look into it some more.

simjanos-dev commented:
Thank you, It looks great! I'll test and merge it when I'll have time.

simjanos-dev commented:
I'll try to test and merge this PR this week.

simjanos-dev committed:
fixed thai

simjanos-dev commented:
Tested it briefly in Norwegian, Japanese and Thai. I added a fix for the last one, but it seems to work well. Thank you so much! This fixes a very annoying issue, a lot of people will appreciate this.
